kimchi/src/tests/xor.rs

```diff
@@ -330,3 +390,81 @@ fn test_bad_xor() {
         ))
     );
 }
+
+#[test]
+// Finalization test
+fn test_xor_finalization() {
+    let num_inputs = 2;
+
+    // circuit
+    let gates = {
+        // public inputs
+        let mut gates = vec![];
+        for row in 0..num_inputs {
+            gates.push(CircuitGate::<Fp>::create_generic_gadget(
+                Wire::for_row(row),
+                GenericGateSpec::Pub,
+                None,
+            ));
+        }
+        // 1 XOR of 128 bits. This will create 8 Xor16 gates and a Generic final gate with all zeros.
+        CircuitGate::<Fp>::extend_xor_gadget(&mut gates, 128);
+        // connect public inputs to the inputs of the XOR
+        gates.connect_cell_pair((0, 0), (2, 0));
+        gates.connect_cell_pair((1, 0), (2, 1));
+
+        // Temporary workaround for lookup-table/domain-size issue
+        for _ in 0..(1 << 13) {
+            gates.push(CircuitGate::zero(Wire::for_row(gates.len())));
+        }
+        gates
+    };
+
+    // witness
+    let witness = {
+        let mut cols: [_; COLUMNS] = array::from_fn(|_col| vec![Fp::zero(); num_inputs]);
+
+        // initialize the 2 inputs
+        let input1 = 0xDC811727DAF22EC15927D6AA275F406Bu128;
+        let input2 = 0xA4F4417AF072DF9016A1EAB458DA80D1u128;
+        cols[0][0] = input1.into();
+        cols[0][1] = input2.into();
+
+        xor::extend_xor_witness::<Fp>(&mut cols, input1.into(), input2.into(), 128);
+        cols
+    };
+
+    let index = {
```

Comment:
This thing feels very repetitive and could be refactored into a function to save some lines of code.  However, not for now.  Maybe after hard-fork in another PR.